src/tests/e2e/pages/metamask.page.js

```diff
@@ -1,27 +1,109 @@
+import pWaitFor from 'p-wait-for'
+import log from 'loglevel'
+log.setDefaultLevel('info')
+
+const POPUP_REGEX = 'chrome-extension://[a-z]+/notification.html'
+const METAMASK_PAGE_REGEX = 'chrome-extension://[a-z]+/home.html'
+const TEST_PAGE = process.env.E2E_TEST_URL
 class MetamaskPage {
-  constructor () {
-    new Promise((resolve, reject) => {
-      browser.on('targetcreated', async target => {
-        if (target.url().match('chrome-extension://')) {
-          try {
-            console.log('found page')
-            const page = await target.page()
-            resolve(page)
-          } catch (e) {
-            reject(e)
-          }
+  async _waitForPopup (open = true) {
+    // wait until popup is open
+    await pWaitFor(async () => {
+      const pages = await browser.pages()
+      const matched = pages.find(p => p.url().match(POPUP_REGEX))
+      return open ? !!matched : !matched
+    })
+  }
+
+  // sometimes, metamask stucks at loading screen
+  // here is a hacky fix by refreshing the popup
+  // see https://github.com/ventureum/Chainsfr/issues/1424
+  // for details
+  async fixHang (metamaskPage) {
+    // check if metamask hangs
+    let spinnerEle
+    do {
+      spinnerEle = await metamaskPage.$('#loading__spinner')
+      if (spinnerEle) {
+        log.info('Looks like metamask hangs, force refresh in popup')
+        await metamaskPage.reload({ waitUntil: ['domcontentloaded'] })
+
+        // recheck
+        await page.waitFor(1000)
+        log.info('rechecking if the popup still hangs...')
+        spinnerEle = await metamaskPage.$('#loading__spinner')
+        if (!spinnerEle) {
+          log.info('Spinner is gone, problem resolved')
+        } else {
+          log.info('Still hangs, re-trying...')
         }
-      })
-    }).then(page => (this.metamaskPage = page))
+      }
+    } while (spinnerEle)
   }
 
-  async approve () {
-    await this.metamaskPage.bringToFront()
-    const confirmButtonSelector =
+  // this is triggered by invoking metamask.enable()
+  async connect () {
+    await this._waitForPopup()
+    const pages = await browser.pages()
+    const metamaskPage = pages.find(p => p.url().match(POPUP_REGEX))
```

Comment:
Have you tried using the getNewPopupPage function from the testUtils.js file?

Reply:
might be unstable for consecutive popups

Reply:
Why do we have consecutive popups? the function returns the first pop up.